### PR TITLE
feat: tasks in projects

### PR DESCRIPTION
**Pull request by djaiss**
image

You fool, don't forget these steps:

- [ ] Unit tests
- [ ] Tests with Cypress
- [ ] Documentation
- [ ] Dummy data

### CODE REVIEW

File: app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        // the trick is to do this with a single query, as we don’t want to do
+        // multiple queries to slow down the loading speed of the page.
+
+        $tasksWithoutLists = $tasks->filter(function ($task) {
+            return is_null($task->project_task_list_id);
+        });
+        $tasksWithoutListsCollection = collect([]);
+        foreach ($tasksWithoutLists as $task) {
+            $tasksWithoutListsCollection->push(self::getTaskInfo($task, $company));
+        }
```
Comment (asbiin) [inline app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php:42]: You can combine the `foreach` with:
```suggestion
        $tasksWithoutListsCollection = $tasksWithoutLists = $tasks->filter(function ($task) {
            return is_null($task->project_task_list_id);
        })->map(function($task) use ($company) {
            return self::getTaskInfo($task, $company));
        })
```

File: app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        foreach ($taskLists as $taskList) {
+            $tasksWithListsCollection = collect([]);
+
+            $tasks = $tasksWithLists->filter(function ($task) use ($taskList) {
+                return $task->project_task_list_id == $taskList->id;
+            });
+
+            foreach ($tasks as $task) {
+                $tasksWithListsCollection->push(self::getTaskInfo($task, $company));
+            }
```
Comment (asbiin) [inline app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php:58]: again:
```suggestion
            $tasksWithListsCollection = $tasksWithLists->filter(function ($task) use ($taskList) {
                return $task->project_task_list_id == $taskList->id;
            })->map(function($task) use ($company) {
                return self::getTaskInfo($task, $company);
            });
```

File: app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+        $employeesCollection = collect([]);
+        foreach ($employees as $employee) {
+            $employeesCollection->push([
+                'value' => $employee->id,
+                'label' => $employee->name,
+            ]);
+        }
+
+        return $employeesCollection;
```
Comment (asbiin) [inline app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php:151]:
```suggestion
        return $project->employees()->notLocked()
            ->get()
            ->map(function($employee) {
                return [
                    'value' => $employee->id,
                    'label' => $employee->name,
                ];
            });
```

File: resources/js/Pages/Project/Tasks/Partials/ProjectTasks.vue [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+          label: task.assignee.name,
+        };
+      }
+
+      // this is really barbaric, but I need to do this to
+      // first: target the TextInput with the right ref attribute
+      // second: target within the component, the refs of the input text
+      // this is because we try here to access $refs from a child component
+      this.$nextTick(() => {
+        this.$refs[`task${task.id}`][0].$refs['input'].focus();
```
Comment (asbiin) [inline resources/js/Pages/Project/Tasks/Partials/ProjectTasks.vue:229]: This is indeed barbaric ...

File: resources/js/Pages/Project/Tasks/Partials/ProjectTasks.vue [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+          flash(this.$t('project.task_create_success'), 'success');
+        })
+        .catch(error => {
+          this.loadingState = null;
+          this.form.errors = error.response.data;
+        });
+    },
+
+    toggle(id) {
+      axios.post(`/${this.$page.props.auth.company.id}/projects/${this.project.id}/tasks/${id}/toggle`)
```
Comment (asbiin) [inline resources/js/Pages/Project/Tasks/Partials/ProjectTasks.vue:257]: toggle should not be a "post" but a "put" instead.

File: routes/web.php [hunk 6]
```diff
@@ -230,6 +230,14 @@
...
             // project messages
             Route::resource('{project}/messages', 'Company\\Project\\ProjectMessagesController', ['as' => 'projects']);
+
+            // project tasks
+            Route::resource('{project}/tasks', 'Company\\Project\\ProjectTasksController', ['as' => 'projects']);
+            Route::post('{project}/tasks/{task}/toggle', 'Company\\Project\\ProjectTasksController@toggle');
+            Route::get('{project}/tasks/lists/create', 'Company\\Project\\ProjectTaskListsController@create');
+            Route::post('{project}/tasks/lists/store', 'Company\\Project\\ProjectTaskListsController@store');
+            Route::put('{project}/tasks/lists/{list}', 'Company\\Project\\ProjectTaskListsController@update');
+            Route::delete('{project}/tasks/lists/{list}', 'Company\\Project\\ProjectTaskListsController@destroy');
```
Comment (asbiin) [inline routes/web.php:240]: this should also be able to be written:
```suggestion
            Route::resource('{project}/tasks/lists', 'Company\\Project\\ProjectTaskListsController')-only([
               'create', 'store', 'update', 'destroy'
            ]);
```

File: sonar-project.properties [hunk 7]
```diff
@@ -7,7 +7,7 @@ sonar.organization=officelife
 sonar.sources=app,bootstrap,config,database,public,resources,routes
 sonar.exclusions=bootstrap/cache/*,public/vendor/**,resources/lang/**
 sonar.tests=tests
-sonar.coverage.exclusions=routes/*.php,config/*.php,bootstrap/**,resources/**/*.php,database/migrations/*.php,public/*.php,resources/**/*.vue,resources/**/*.js,**/*Controller.php
+sonar.coverage.exclusions=routes/*.php,config/*.php,bootstrap/**,resources/**/*.php,database/migrations/*.php,public/*.php,resources/**/*.vue,resources/**/*.js,**/*Controller.php,database/factories/*
```
Comment (asbiin) [inline sonar-project.properties:10]:
```suggestion
sonar.coverage.exclusions=routes/*.php,config/*.php,bootstrap/**,resources/**/*.php,database/factories/*,database/migrations/*.php,public/*.php,resources/**/*.vue,resources/**/*.js,**/*Controller.php
```

File: webpack.mix.js [hunk 8]
```diff
@@ -5,7 +5,7 @@ require('laravel-mix-purgecss');
 let purgeCssOptions = {
   enabled: true,
   // List of regex of CSS class to not remove
-  whitelistPatterns: [/^ball-pulse/, /^vs__/],
+  whitelistPatterns: [/^ball-pulse/, /^vs__/, /dot-/],
```
Comment (asbiin) [inline webpack.mix.js:8]: `^` means the beginning of the pattern
```suggestion
  whitelistPatterns: [/^ball-pulse/, /^vs__/, /^dot-/],
```

File: README.md [hunk 9]
```diff
@@ -1,13 +1,12 @@
 [![Build and test](https://github.com/officelifehq/officelife/workflows/Build%20and%20test/badge.svg)](https://github.com/officelifehq/officelife/actions?query=workflow%3A%22Build+and+test%22)
-[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=officelife&amp;metric=ncloc)](https://sonarcloud.io/dashboard?id=officelife)
 [![Code coverage](https://sonarcloud.io/api/project_badges/measure?project=officelife&amp;metric=coverage)](https://sonarcloud.io/project/activity?custom_metrics=coverage&amp;graph=custom&amp;id=officelife)
 [![Quality gate](https://sonarcloud.io/api/project_badges/measure?project=officelife&amp;metric=alert_status)](https://sonarcloud.io/dashboard?id=officelife)
-[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)
+[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=officelife&amp;metric=ncloc)](https://sonarcloud.io/dashboard?id=officelife)
```
Comment (asbiin) [inline README.md:4]: This should be moved in a dedcated PR

File: app/Console/Commands/Tests/SetupDummyAccount.php [hunk 10]
```diff
@@ -242,6 +245,9 @@ private function createFirstUser(): void
             'last_name' => 'Scott',
         ]);
 
+        $user->email_verified_at = Carbon::now();
```
Comment (asbiin) [inline app/Console/Commands/Tests/SetupDummyAccount.php:248]: Right, I've missed that one. you can also set  `MAIL_VERIFY=false` in you `.env` file

**Review by asbiin — CHANGES_REQUESTED** [review]
That's a big one!
I did not try it yet, I willl definitely enjoy it 👍

File: app/Http/Controllers/Company/Project/ProjectTaskListsController.php [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+            'project_task_list_id' => $projectTaskListId,
+            'title' => $request->input('title'),
+            'description' => $request->input('description'),
+        ];
+
+        $taskList = (new UpdateProjectTaskList)->execute($data);
+
+        return response()->json([
+            'data' => ProjectTasksViewHelper::getTaskListInfo($taskList),
+        ], 201);
```
[the remark this answers is not in the document]
Comment (djaiss) [reply app/Http/Controllers/Company/Project/ProjectTaskListsController.php:104]: Good one.

File: resources/js/Pages/Project/Tasks/Partials/ProjectTasks.vue (hunk 4 above)
Comment (djaiss) [reply resources/js/Pages/Project/Tasks/Partials/ProjectTasks.vue:229]: I know :-(

File: resources/js/Pages/Project/Tasks/Partials/ProjectTasks.vue (hunk 5 above)
Comment (djaiss) [reply resources/js/Pages/Project/Tasks/Partials/ProjectTasks.vue:257]: 👍

File: routes/web.php (hunk 6 above)
Comment (djaiss) [reply routes/web.php:240]: Done!

File: app/Console/Commands/Tests/SetupDummyAccount.php (hunk 10 above)
Comment (djaiss) [reply app/Console/Commands/Tests/SetupDummyAccount.php:248]: OK, i’ll also do it in a separate PR.

File: README.md (hunk 9 above)
Comment (djaiss) [reply README.md:4]: Done

File: app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php (hunk 1 above)
Comment (djaiss) [reply app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php:42]: It’s almost ok without `$tasksWithoutLists =` in your suggestion.

Comment (djaiss) [reply app/Http/ViewHelpers/Project/ProjectTasksViewHelper.php:42]: Or... no I’m wrong.

**Comment by djaiss** [comment]
@asbiin do you know how to fix the phpstan error on this PR?

**Comment by asbiin** [comment]
@djaiss I guess the best way is to "cast" with an intermediate variable:
```php
        /** @var \Database\Factories\Company\ProjectTaskFactory */
        $ptf = ProjectTask::factory();
        $projectTaskA = $ptf->completed()->create([
```
... but it would be ugly to have to do it everywhere.

So I guess we'll exclude this error in the `phpstan.neon` file:
```
 - '#Call to an undefined method Illuminate\\Database\\Eloquent\\Factories\\Factory::[a-zA-Z0-9\\_]+\(\)\.#'
```

**Comment by djaiss** [comment]
@asbiin thanks so much!

**Pull request merged**